src/firewall/core/fw_config.py

```diff
@@ -174,6 +174,109 @@ def update_firewalld_conf(self):
         else:
             self._firewalld_conf.read()
 
+    # helper funtion (for detecting if builtin outside this module without throwing an error)
+
+    def is_builtin(self, obj):
+        if obj.builtin or not obj.default:
+            return True
+        return False
+
+    # Reset defaults
+
+    def new_io_object(self, obj, conf):
+        obj_type = type(obj)
+        name = obj.name
+        log.debug1("new %s - %s" % (str(obj_type), obj.name))
+        if obj_type is IPSet:
+            self.new_ipset(name, conf)
+        elif obj_type is Helper:
+            self.new_helper(name, conf)
+        elif obj_type is IcmpType:
+            self.new_icmptype(name, conf)
+        elif obj_type is Service:
+            self.new_service(name, conf)
+        elif obj_type is Zone:
+            self.new_zone(name, conf)
+        elif obj_type is Policy:
+            self.new_zone(name, conf)
+        else:
+            raise FirewallError(errors.INVALID_OBJECT,
+            "obj named %s of type %s passed into fw_config.new_io_object()" % (obj.name, str(type(obj))))
+
+    def set_io_object(self, obj, conf):
+        obj_type = type(obj)
+        log.debug1("setting %s - %s" % (str(obj_type), obj.name))
+        if obj_type is IPSet:
+            self.set_ipset_config(obj, conf)
+        elif obj_type is Helper:
+            self.set_helper_config(obj, conf)
+        elif obj_type is IcmpType:
+            self.set_icmptype_config(obj, conf)
+        elif obj_type is Service:
+            self.set_service_config(obj, conf)
+        elif obj_type is Zone:
+            self.set_zone_config(obj, conf)
+        elif obj_type is Policy:
+            self.set_policy_object_config_dict(obj, conf)
+        else:
+            raise FirewallError(errors.INVALID_OBJECT,
+            "obj named %s of type %s passed into fw_config.set_io_object()" % (obj.name, str(type(obj))))
+
+    def remove_io_object(self, obj):
+        obj_type = type(obj)
+        log.debug1("removing/resetting %s - %s" % (str(obj_type), obj.name))
+        if obj_type is IPSet:
+            self._remove_ipset(obj)
+        elif obj_type is Helper:
+            self._remove_helper(obj)
+        elif obj_type is IcmpType:
+            self._remove_icmptype(obj)
+        elif obj_type is Service:
+            self._remove_service(obj)
+        elif obj_type is Zone:
+            self._remove_zone(obj)
+        elif obj_type is Policy:
+            self._remove_policy_object(obj)
+        else:
+            raise FirewallError(errors.INVALID_OBJECT,
+            "obj named %s of type %s passed into fw_config.remove_io_object()" % (obj.name, str(type(obj))))
+
+    def _reset_defaults(self):
+        order = [self._zones, self._policy_objects, self._ipsets, self._services, self._helpers, self._icmptypes]
+        for setting in order:
+            setting_copy = copy.copy(setting)
+            for obj in setting_copy.values():
+                self.remove_io_object(obj)
```

Comment:
It's a lot of work to sort out these io_objects and find the correct remove function. Since we're going to reload can we simply remove/unlink the XML files in `/etc/firewalld/{zones,policies,services,ipsets,icmptypes,helpers,etc}`?

Reply:
I have now updated it to just remove the xml files in the same way as the type specific methods would. Working to find a good workaround for {set,new}_io_object() methods for restoring settings in the event of reset's failure.